provides.py

```diff
@@ -131,6 +131,10 @@ def changed(self):
             return False
         saved_hash = unitdata.kv().get(self._hash_key)
         result = saved_hash != self.hash
+        if not result and saved_hash and not self._is_completed:
+            # Mark as completed all requests, which come from the same EC2
+            # and require the same set of permissions and permissions were already granted.
+            self.mark_completed()
```

Comment:
Following how this fixes the problem took me a while. Maybe have the code express more directly the condition we're trying to fix?

We want to mark completed a request from one unit matching another already fulfilled for another unit on the same instance. Maybe something like this (renaming `IntegrationRequest._is_completed` to `IntegrationRequest.completed_for_instance`)?
``` python
    @when('endpoint.{endpoint_name}.changed')
    def check_requests(self):
        unfulfilled_requests = False
        for request in self.all_requests:
            if request.changed:
                # request differs from anything previously completed for this instance
                unfulfilled_requests = True
            elif request.completed_for_instance:
                # request completed for a different unit on the same instance
                request.mark_completed()
        toggle_flag(self.expand_name('requested'), unfulfilled_requests)
        clear_flag(self.expand_name('changed'))
```
I also like that, with this version, a property getter doesn't change state elsewhere. While I expect reading a property might update a variable that backs it (e.g., storing the result of an external call), I don't expect a side effect in a separate data structure.

Reply:
Thanks for your suggestions. Updated.